psycopg3/connection.py

```diff
@@ -85,6 +94,9 @@ def __init__(self, pgconn: pq.proto.PGconn):
         self._notify_handlers: List[NotifyHandler] = []
         # name of the postgres encoding (in bytes)
         self._pgenc = b""
+        # stack of savepoint names managed by active Transaction() blocks
+        # (The first item may be None, representing an outer transaction)
+        self._savepoints: List[Union[None, str]] = []
```

Comment:
I'm not thrilled with this being a list of savepoint names or, optionally, for the first item only, a marker (currently `None`) indicating top-level transaction. 

However I can't think of a better way to give Connection a way to know if any Transaction contexts are active (so that it can deny access to commit/rollback/autocommit) -- currently the criteria is if `self._savepoints` is non-empty, then transaction contexts are active.

It's worth noting that entering a transaction context may:
1. Begin an outer transaction (if one isn't already in progress)
2. Begin an outer transaction _and_ create a savepoint (if one is named -- though names aren't implemented yet)
3. Create a savepoint (if a transaction is already in progress)

So there isn't even a 1-to-1 correspondence between items in this list and Transaction instances, because in case 2 a single Transaction instance may result in two items in this list.

(This complexity is part of the reason that we decided to have psycopg-nestedtransactions _always_ create a savepoint, because it then takes away case 1, and therefore you don't need to have a None marker-item in the list because you'll get a savepoint there to make it non-empty. BUT, we've said that the overhead of creating a savepoint every time you begin a transaction is unnecessary overhead that is not appropriate, and so we find ourselves here...)

Reply:
It's only a placeholder. You could use an empty string or a string symbolic constant. Or have an `Optional[List[str]]` and give different meaning to None and empty list. I also feel that maybe we should use `pgconn.transaction_status` to detect if we are inside or outside the first level transaction.

Reply:
I like `Optional[List[str]]`. None means no active scopes; non-none is a list of active savepoints (and the empty list is a perfect fit for "no active savepoints").

Bonus 1: Will also give me an excuse to check for `None` explicitly rather than relying on truthiness (a piece of PEP 8 I personally disagree with!) :-)
Bonus 2: Allows easier auto-naming of savepoints. (I see you already spotted that bit of cumbersome logic.)

Reply:
I am very ok for explicitly checking for None. `is None` also happens to be a particularly fast operation, as it doesn't drop into the whole rich comparison machinery.